tests/test_git_utils.py

```diff
@@ -1,4 +1,842 @@
+
+        with pytest.raises(GitCliError):
+            cli.reset_hard(dir_path)
+
+        with pytest.raises(GitCliError):
+            cli.add_remote(dir_path, "origin", "https://github.com/torvalds/linux.git")
+
+        with pytest.raises(GitCliError):
+            cli.fetch_all(dir_path)
+
+        assert not cli.does_branch_exist(dir_path, "main")
+
+        with pytest.raises(GitCliError):
+            cli.checkout_branch(dir_path, "main")
+
+
+# noinspection PyProtectedMember
+def init_temp_git_repo(git_dir: Path):
+    cli = GitCli()
+    cli._run_git_command(["init"], working_directory=git_dir)
+    cli._run_git_command(
+        ["config", "user.name", "CI Test User"], working_directory=git_dir
+    )
+    cli._run_git_command(
+        ["config", "user.email", "ci-test-user-invalid@example.com"],
+        working_directory=git_dir,
+    )
+
+
+def test_git_cli_reset_hard_already_reset():
+    cli = GitCli()
+    with tempfile.TemporaryDirectory() as tmpdir:
+        dir_path = Path(tmpdir)
+
+        init_temp_git_repo(dir_path)
+        cli._run_git_command(
+            ["commit", "--allow-empty", "-m", "Initial commit"],
+            working_directory=dir_path,
+        )
+
+        cli.reset_hard(dir_path)
+
+
+@mock.patch("conda_forge_tick.git_utils.GitCli._run_git_command")
+def test_git_cli_reset_hard_mock(run_git_command_mock: MagicMock):
+    cli = GitCli()
+
+    git_dir = Path("TEST_DIR")
+
+    cli.reset_hard(git_dir)
+
+    run_git_command_mock.assert_called_once_with(
+        ["reset", "--quiet", "--hard", "HEAD"], git_dir
+    )
+
+
+def test_git_cli_reset_hard():
+    cli = GitCli()
+    with tempfile.TemporaryDirectory() as tmpdir:
+        dir_path = Path(tmpdir)
+
+        init_temp_git_repo(dir_path)
+        cli._run_git_command(
+            ["commit", "--allow-empty", "-m", "Initial commit"],
+            working_directory=dir_path,
+        )
+
+        with dir_path.joinpath("test.txt").open("w") as f:
+            f.write("Hello, World!")
+
+        cli._run_git_command(["add", "test.txt"], working_directory=dir_path)
+        cli._run_git_command(
+            ["commit", "-am", "Add test.txt"], working_directory=dir_path
+        )
+
+        with dir_path.joinpath("test.txt").open("w") as f:
+            f.write("Hello, World! Again!")
+
+        cli.reset_hard(dir_path)
+
+        with dir_path.joinpath("test.txt").open("r") as f:
+            assert f.read() == "Hello, World!"
+
+
+def test_git_cli_clone_repo_not_exists():
+    cli = GitCli()
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        dir_path = Path(tmpdir)
+
+        with pytest.raises(GitCliError):
+            cli.clone_repo(
+                "https://github.com/conda-forge/this-repo-does-not-exist.git", dir_path
+            )
+
+
+def test_git_cli_clone_repo_success():
+    cli = GitCli()
+
+    git_url = "https://github.com/conda-forge/duckdb-feedstock.git"
```

Comment:
I am pretty sure this feedstock has been archived based on other PRs I looked at today. Maybe pick another one? Sometimes I used conda-smithy or similar since we likely won't ever archive that one.

Reply:
Actually I used this feedstock exactly because it has been archived. My criteria for a testing feedstock was:

1. The feedstock should not be deleted in the near future.
2. Its content should somehow be consistent, i.e. the README file should not disappear because it is used in some tests.
3. It should have as little commits as possible to make testing faster.

conda-smithy and other popular feedstocks satisfy all of these except 3. Less popular feedstocks might violate 2 because we don't really have control about what happens in these.

This is why I chose an archived feedstock: Its content will not change, and also the number of commits is fixed. I did this with the assumption that conda-forge does not delete archived feedstocks - what do you think?

Reply:
Oh I didn't realize github let you clone archived repos.